app/Fission/Internal/Orphanage/RIO.hs

```diff
@@ -3,20 +3,64 @@
 
 module Fission.Internal.Orphanage.RIO () where
 
+import Fission.Prelude
+
 import Data.Pool
 import RIO.Orphans ()
 
 import Database.Selda.Backend.Internal
 import Database.Selda.PostgreSQL
 
-import Fission.Prelude
+import qualified RIO.ByteString.Lazy as Lazy
+
+import           Servant.Client
+import           Network.HTTP.Client as HTTP
 
-import qualified Fission.Config as Config
+import qualified Fission.Config        as Config
 import qualified Fission.Storage.Types as DB
 
+import           Network.IPFS
+import           Network.IPFS.Types         as IPFS
+import qualified Network.IPFS.Process.Error as Process
+import           Network.IPFS.Process
+
 instance Has (DB.Pool PG) cfg => MonadSelda (RIO cfg) where
   type Backend (RIO cfg) = PG
 
   withConnection action = do
     DB.Pool pool <- Config.get
     withResource pool action
+
+instance 
+  ( HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => MonadLocalIPFS (RIO cfg) where
+    ipfsRun opts arg = do
+      IPFS.BinPath ipfs <- Config.get
+      IPFS.Timeout secs <- Config.get
+      let opts' = ("--timeout=" <> show secs <> "s") : opts
+
+      runProc readProcess ipfs (byteStringInput arg) byteStringOutput opts' >>= \case
+        (ExitSuccess, contents, _) ->
+          return <| Right contents
+        (ExitFailure _, _, stdErr)
+          | Lazy.isSuffixOf "context deadline exceeded" stdErr -> 
+              return . Left <| Process.Timeout secs
+          | otherwise ->
+            return . Left <| Process.UnknownErr stdErr
+
+instance 
+  ( Has IPFS.URL cfg
+  , Has HTTP.Manager cfg
+  )
+  => MonadRemoteIPFS (RIO cfg) where
+    runRemote query = do
+      IPFS.URL url            <- Config.get
+      manager :: HTTP.Manager <- Config.get
+      url
+        |> mkClientEnv manager
+        |> runClientM query
+        |> liftIO
```

Comment:
🤔 Any thoughts about doing a swarm connect before every action right in the instance? Just port in the kludge from the CLI and automate away the details so we don't have to look at the hack and feel bad 🤣

Reply:
:thinking: :thinking: 
We wouldn't want that in the web-api would we? I dig the idea for in the cli when we write these instances tho

Reply:
Aren't we still having issues where the web-api randomly can't find the S3 node? The reconnection script may not be frequent enough or... something? I dunno. I had this flagged in the back of my brain for a few weeks, so that info may be stale. Thoughts?

Reply:
ah right that's true. yup I'll go ahead and do it in both then

Reply:
Tagging so this doesn't get lost (@expede)

So in order to connect, you need to be in a `MonadLocalIPFS`. I'm not sure we want to attach that as a constraint on our `MonadRemoteIPFS` instance. Thoughts?

Reply:
That's a good point 🤔 You could do something along these lines:

```haskell
class Monad m => MonadRemoteIPFS m where
  runRemote :: ClientM a -> m(Either ClientError a)
  ipfsConnect :: Peer -> m (Either ConnectionError ())
  ipfsAdd   :: Lazy.ByteString -> m (Either ClientError CID)
  -- and so on ...
  ipfsAdd raw = runRemote <| IPFS.Client.add raw
  ipfsConnect _ = return () -- Noop
  -- and the rest of the defaults

instance 
  ( Has IPFS.URL     cfg
  , Has HTTP.Manager cfg
  , Has Peer         cfg
  , MonadRIO         cfg m
  , MonadLocalIPFS m
  ) 
  => MonadRemoteIPFS m where
    runRemote = do
      Peer peerID             <- Config.get
      IPFS.URL url            <- Config.get
      manager :: HTTP.Manager <- Config.get

      Local.connectPeer peerID

      url
        |> mkClientEnv manager
        |> runClientM query
        |> liftIO
```